### PR TITLE
Add exact calculation functions for nxm/eth price & spot price & price impact formulas

### LINKED ISSUE #69: Tokenomics price formulas in the SDK
- Calculate price (this includes price impact):
  - swap ETH -> NXM (calculateNXMForETH(ethInput)) 
      ```
      k = state.nxmA * state.eth
      ethNew = state.eth + ethInput
      nxmANew = k / ethNew
      nxmOut = nxmANew - state.nxmA
      ```
      
  - swap ETH -> NXM (calculateETHForExactNXM(nxmOutput)) 
    ```
     k = state.nxmA * state.eth
     nxmANew = nxmA - nxmOutput
     ethNew = k / nxmANew
     ethInput = ethNew - state.Eth
    ```
    
  - swap NXM -> ETH (calculateETHForNXM nxmInput))
      ```
      k = state.nxmB * state.eth
      nxmBNew = state.nxmB + nxmInput
      ethNew = k / nxmBNewethNew
      ethOut = state.eth - ethNew
      ```
      
  - swap NXM -> ETH (calculateNXMForExactETH(ethOutput)) 
      ```
      k = state.nxmB * state.eth
      ethNew = state.eth - ethOutput
      nxmBNew = k / ethNew
      nxmIn = nxmBNew - state.nxmB
      ```
      
- Calculate spot prices (don't include price impact):
    ```
        State state = getReserves();
        return (
          1 ether * state.eth / state.nxmA,  // spot price A
          1 ether * state.eth / state.nxmB  // spot price B
        );
    ```
  
- Calculate price impact (expressed as a %): 
  - swap ETH -> NXM (buy NXM)  
     ```
     priceImpactA = abs (1 - calculateNXMForETH() / spotPriceA) * 100
  - swap NXM -> ETH (sell NXM)  
     ```
     priceImpactB = abs (1 - calculateETHForNXM() / spotPriceB) * 100
     ```

### PR DESCRIPTION
**Pull request by Gideonnn**
Small fixes and added unit tests.

Closes #69

### CODE REVIEW

File: src/swap/calculateEthForNxm.ts [hunk 1]
```diff
@@ -1,19 +1,19 @@
 import { Reserves } from './reserves.type';
 
+/**
+ * Calculates the amount of eth flowing out of the pool for the given amount of nxm flowing in
+ */
+// suggestion: calculateEthOutGivenNxmIn
 export const calculateEthForNxm = (nxmIn: bigint, reserves: Reserves): bigint => {
```
Comment (Gideonnn) [inline src/swap/calculateEthForNxm.ts:7]: What do you think about renaming this function to the suggestion above, or something else more verbose? All four functions have a suggestion above it

File: package.json [hunk 2]
```diff
@@ -46,6 +46,7 @@
     "semver": "^7.4.0",
     "ts-jest": "^29.1.1",
     "tsup": "^7.1.0",
-    "typescript": "^5.1.6"
+    "typescript": "^5.1.6",
+    "viem": "^1.16.6"
```
Comment (Gideonnn) [inline package.json:50]: Added `viem` mostly for the `parseEther` and other utilities, we will probably use them more often in unit tests

File: src/swap/calculateEthForExactNxm.test.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    nxmA: 142858457219554100789497n,
+    nxmB: 328817222320643252285179n,
+    ethReserve: 5000000000000000000000n,
+    budget: 43835000000000000000000n,
+  };
+
+  const cases = [
+    ['unit value', '1', 34999922981378727n],
+    ['decimal value', '0.1', 3499970248373829n],
+    ['large value', '1000000', 35006677919768798199868n], // ! this returns a negative value - why?
```
Comment (mixplore) [inline src/swap/calculateEthForExactNxm.test.ts:17]: ⚠️ Tests fail for large values like `1,000,000` - the method returns a negative value; why?

File: src/swap/calculateEthForExactNxm.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { Reserves } from './reserves.type';
+
+/**
+ * Calculates the amount of eth needed to flow into the pool to get the exact amount of nxm out
+ */
+// suggestion: calculateEthInGivenNxmOut
+export const calculateEthForExactNxm = (nxmOut: bigint, reserves: Reserves): bigint => {
+
+  if(nxmOut <= 0n) throw new Error('NXM out value must be greater than 0');
```
Comment (shark0der) [inline src/swap/calculateEthForExactNxm.ts:9]: I think `if(nxmOut > reserves.nxmA) { ... }` should be there as well and would prevent the negative `ethIn`

Comment (mixplore) [reply src/swap/calculateEthForExactNxm.ts:9]: I assume for the other exact method would be `ethOut > reserves.ethReserve`. I'll add the conditions for the others as well 👍

**Review by mixplore — APPROVED** [review]

**Comment by mixplore** [comment]
Will merge this now for easier integration with the WIP app changes. Will address any feedback from this PR in a separate one resulting in a new package deploy  👍

**Pull request merged**